### DISCUSSION TITLE
Encrypted .env to remove password exposure?

### DISCUSSION THREAD

**Discussion by queejie** [prisma/prisma#8454] [Q&A]
We're looking at switching over from TypeOrm.  We currently use an encrypted .env file for our app so our connection information is not exposed. Even a db user with minimal privileges can select on sensitive information.  It doesn't appear that there is a way to hide the password in the connection string.  If we wanted to run `prisma introspect`, for example, we'd have to set an environment variable manually every time we ran the command, which would expose the password in our command line history.  Is there a way to encrypt the .env file for prisma, or have it retrieve some information from an encrypted file?  Thank you in advance.

**Comment by ryands17** [comment]
@queejie 👋 

Unfortunately we do not support encrypted env variables at this time for CLI commands. It would be great if you could create a feature request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that we can look into this 🙂 

As a workaround, one way to do this would be running your decryption script before the command so that only the command gets the correct env variable and it's not stored in your command history. So something like:

```
node -r decryptEnvCommonJSModule ./node_modules/.bin/prisma introspect
```

Could you try this and let us know if this works?

**Reply by ryands17** [reply]
Also how are you doing it with TypeORM using the encrypted `.env`?

**Reply by queejie** [reply]
Thank you for the suggestions!  I will try the workaround and get back to you. I'm not super adept with node yet, so it may take a little bit of time.  I suspect it will work, but may rely on global environment variables rather than just using ones scoped to the prisma process alone.  I'm not sure if that's a big deal, security wise.

Regarding how it works with TypeOrm, we're using the secure-env package for our app, which I believe creates environment variables scoped just to the app process.  But it hasn't been solved for the entity/model generation, which is what led me (indirectly) to start looking for alternatives.  The typeorm-model-generator relies on a clear password on the command line.  I guess that's not a big deal, since it isn't stored in files on disk, and I can rely on using a SQL user with SELECT only privileges.

The larger question about the future of TypeORM was raised by the author of typeorm-model-generator, however, who seems to be saying that he is done with development on that tool.  He doesn't see much of a future in TypeORM.  I've been trying to understand if that is a valid perspective.  I've read the docs page here on TypeORM vs Prisma, and found it very helpful.  I've been very happy with the TypeORM paradigm and software, finding nothing lacking.  We've used it for a number of very complex applications without many issues.  

The only real issue I have with TypeORM is that the entity classes generated are difficult, if not impossible, to share with the client side Angular app.  The shim doesn't seem to work for us, and I've ended up creating nearly duplicate 'record' files for the client, cleaned of decorators.  The ideal for me would be a set of entities that is shared between the front and back ends of the app, all stemming from a well-designed SQL schema.

**Comment by queejie** [comment]
To answer more specifically how we use secure-env, we do this:

```openssl rand -hex 64 > .env.key && secure-env .env -s `cat .env.key` && rm -f .env```

to encrypt, and used these lines to decrypt, in the app:

````
const basePath = join(__dirname, '../../', `.env`)
const secret = readFileSync(`${basePath}.key`, {encoding: 'utf8'}).trim();
global.env = secureEnv({secret, path: `${basePath}.enc`});
````

So the key is stored with the encrypted env file.  We don't store either in git, of course.  Not perfect, but I think it does provide some more protection.

**Reply by janpio** [reply]
I came here (after an internal discussion) to ask exactly these details - thanks for anticipating :D 

This then really just protects against some malicious npm package that picks your `.env` file and uploads it to a server - but if that package also reads the other file, or injects into your code and then reads `global.env` after this was executed, the connection string is equally available for them, correct?

This is an interesting use case - not sure if it is common enough though to implement this in Prisma Client itself vs. e.g. doing exactly the code you posted above in your application, that uses Prisma.

**Reply by queejie** [reply]
Yeah, I guess it's just a high level of obfuscation if it is grabbed by a malicious npm package. I need to read up on the npm audit feature to see if it checks for that sort of thing.  We use different env files and keys for every machine, whether dev, staging or production.  But if it can be grabbed at runtime I can see the problem.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
